### PR TITLE
dev - identifier implemented and rebased

### PR DESCRIPTION
**Pull request by CCranney**

### CODE REVIEW

File: csodiaq/identifier/outputWritingFunctions.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+def extract_metadata_from_match_dataframe_groupby(group):
+    return {
+        "shared": len(group.index),
+        "ionCount": sum(group["queryIntensity"]),
+        #TODO: see TODO in extract_metadata_from_match_and_score_dataframes test
+        #"extractNum": 0
+    }
+
+def extract_metadata_from_score_dataframe(df):
+    #TODO: there has to be a cleaner way to do this
```
Comment (CCranney) [inline csodiaq/identifier/outputWritingFunctions.py]: In a nutshell, I want to make a dictionary with the following shape:

key: (libraryIdx, queryIdx)
value: {
   'maccScore': X,
   'cosineScore':X,
}

Each row is unique, so used the pandas to_dict to convert the scoreDf (with columns 'libraryIdx','queryIdx','maccScore','cosineScore') to basically make 2 dictionaries corresponding to maccScore and cosineScore, respectively, before smooshing them together into one dictionary.

I may make a stack overflow comment on this - I wasn't finding anything that answered this specific question.

File: csodiaq/csodiaq.py [hunk 2]
```diff
@@ -27,16 +28,18 @@ def main():
         return
 
     if args['command'] == 'id':
-        lib = cif.library_file_to_dict(args['library'])
-        maxQuerySpectraToPool = queryPooling = args['query']
-        if not maxQuerySpectraToPool:
-            maxQuerySpectraToPool = np.inf
+        identifier = Identifier(args)
+        identifier.identify_library_spectra_in_queries()
+        #lib = cif.library_file_to_dict(args['library'])
```
Comment (CCranney) [inline csodiaq/csodiaq.py:44]: I'm keeping this commented code around for comparison with the old code as I continue to rework it.

File: csodiaq/identifier/outputWritingFunctions.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        matchMetadata["cosineSimilarityScore"],
+        libMetadata["identifier"],
+        queMetadata["peaksCount"],
+        libMetadata["peaks"],
+        matchMetadata["shared"],
+        matchMetadata["ionCount"],
+        queMetadata["CV"],
+        queMetadata["windowWidth"],
+        matchMetadata["maccScore"],
+        #matchMetadata["excludeNum"],
```
Comment (CCranney) [inline csodiaq/identifier/outputWritingFunctions.py:20]: I've made a TODO about this and will be asking Jesse about it. I expect this to be re-added to the output once I've parsed that out.

File: csodiaq/identifier/scoringFunctions.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+def identify_nearest_bin_to_tallest_bin_that_represents_noise(binHeights, tallestBinIdx):
+    averageNoise = np.mean(binHeights[:10] + binHeights[-10:])
+    binsBeforeTallest = binHeights[:tallestBinIdx]
+    beforeNearestNoiseBinIdx = len(binsBeforeTallest) - np.argmin(binsBeforeTallest[::-1] >= averageNoise) - 1
+    binsAfterTallest = binHeights[tallestBinIdx:]
+    afterNearestNoiseBinIdx = np.argmin(binsAfterTallest >= averageNoise) + tallestBinIdx
+    nearestNoiseBinIdx = min([beforeNearestNoiseBinIdx, afterNearestNoiseBinIdx], key=lambda x: abs(tallestBinIdx - x))
+    return nearestNoiseBinIdx
+
+def create_ppm_histogram(ppms, offset, tolerance):
```
Comment (CCranney) [inline csodiaq/identifier/scoringFunctions.py:70]: The plot generation code works, but has not been integrated into the main workflow yet. I've added it to my TODO tasks.

File: csodiaq/csodiaq.py [hunk 5]
```diff
@@ -12,6 +12,7 @@
 from . import csodiaq_quantification_functions as cqf
 from . import csodiaq_mgf_cleaning_functions as cmf
 from . import peptide_quantification
+from csodiaq.identifier.Identifier import Identifier
```
Comment (AlexandreHutton) [inline csodiaq/csodiaq.py]: By convention, modules (e.g. `Identifier.py`) are lower-case. If you rename `Identifier.py` to `identifier.py` and add the following line to `identifier/__init__.py`:  
`from .identifier import Identifier`  
You should be able to use `from csodiaq.identifier import Identifier`.

File: csodiaq/csodiaq.py (hunk 2 above)
Comment (AlexandreHutton) [reply csodiaq/csodiaq.py:44]: Alternatively, you could look at the old version by looking at the relevant commit on GitHub, or view the previous commit directly in Pycharm (`Alt` + `9` -> find commit -> right-click -> compare with local).  
It's not as quick as having the commented code there, I realize, but it lets you get rid of unneeded code.

File: csodiaq/identifier/Identifier.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    Class for identifying library peptides in input query spectra for mass spec DIA experiments.
+
+    Extended Summary
+    ----------------
+    The purpose of this class is to consolidate the primary workflow of identifying peptides
+        from a library file in mass spec DIA query files. This includes finding similar peaks
+        between them both (matching) and eliminating false positive matches (scoring). There is
+        an optional correction step that further eliminates false positive library-query peak
+        matches, refining both the final scores and false positive peptide elimination.
```
Comment (AlexandreHutton) [inline csodiaq/identifier/Identifier.py]: This is a good explanation.

File: csodiaq/identifier/Identifier.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    _decoySet : set
+        A set containing all the keys of self._libraryDict that represent a decoy insert.
+            Decoys are used to calculate the probability that a non-decoy match is a
+            false positive.
+    """
+
+    def __init__(self, args):
+        self._args = args
+        self._libraryDict = LibraryLoaderContext(self._args["library"]).load_csodiaq_library_dict()
+        self._decoySet = set([key for key,value in self._libraryDict.items() if value["isDecoy"]])
```
Comment (AlexandreHutton) [inline csodiaq/identifier/Identifier.py]: It looks like you're expecting users to gather their inputs into a dictionary and send them to `Identifier`. There are times where you might want to do that, but I don't think that this is such a time. If it is, that's fine, but there should be some description of what the dictionary is expected to contain; I can't tell from the `__init__` function what `args` should look like. At minimum, pointing to where the argument list might have been generated if it's large/complicated: "the dictionary as generated by function/process XYZ".  

I think this is intended to be `**kwargs` instead of `args`. The latter allows for only one positional argument; the latter allows for many.  
`args` here signals a single, positional argument. `buf = Identifier(1,2)` would throw an error, saying that only one positional argument is needed.  
`*args` would signal that the function accepts any number of positional arguments. `Identifier(1,2)` would work fine, but `Identifier(1, library=3)` would not, since the function doesn't support kwargs.  
`**kwargs` gathers all kwargs into a dictionary. So here, you'd have `Identifier(library=lib, other_input=5, etc="etc")`.  
  
I'm guessing that the parts that call this pass in a dictionary; if you decide to update the signature, keep in mind that you can use `Identifier(**the_dict_you_were_using)` to expand the dictionary into a keyword list.

File: csodiaq/identifier/Identifier.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        """
+        The primary function called for matching library spectra to query spectra.
+            This is the only public-facing function of the class.
+        """
+        for queryFile in self._args["files"]:
+            self._queryContext = QueryLoaderContext(queryFile)
+            matchDf = self._match_library_to_query_spectra()
+            scoreDf = self._score_spectra_matches(matchDf)
+            matchDf, scoreDf = self._apply_correction_to_dataframes(matchDf, scoreDf)
+            self._write_identifications_to_dataframe(matchDf, scoreDf)
```
Comment (AlexandreHutton) [inline csodiaq/identifier/Identifier.py]: This does identification + writes out the data to a file that I didn't specify. I would not expect the identification function to potentially overwrite a file.

File: csodiaq/identifier/Identifier.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+    def _match_library_to_query_spectra(self):
+        """
+        This function identifies peaks in library and query spectra that are within a similar
+            m/z value. This function includes an initial filtering step that removes all
+            library-query spectrum matches that contain fewer than 3 peak matches.
```
Comment (AlexandreHutton) [inline csodiaq/identifier/Identifier.py]: It should be possible to have the number of peaks as a parameter. 
The [filtering function](https://github.com/xomicsdatascience/CsoDIAq/blob/ef1fd0a9c5f9964ced4232a8fca86a22aa5e4cc3/csodiaq/identifier/matchingFunctions.py#L36) can be changed to have `minNumMatches = 3` in the argument list instead.

File: csodiaq/identifier/Identifier.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        matchDf : pandas DataFrame
+            A filtered version of the matchDf input parameter, where peak matches with a ppm value
+                outside the calculated expected range are removed. The removal of all spectral matches
+                with fewer than 3 peak matches is repeated as well.
+
+        scoreDf : pandas DataFrame
+            The score dataframe is recalculated from the ground up using the filtered match dataframe.
+        """
+        if self._args["correction"] == -1:
+            return matchDf, scoreDf
```
Comment (AlexandreHutton) [inline csodiaq/identifier/Identifier.py]: This check should be done before calling this method; if the method is called, it should be assumed that it was done intentionally.

File: csodiaq/identifier/Identifier.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        for key, matchMetadata in matchDict.items():
+            libKeyIdx, queryScan = key
+            libraryMetadata = self._prepare_library_dictionary_for_output(libKeyIdx, sortedLibKeys)
+            queryMetadata = queryDict[str(queryScan)]
+            queryMetadata["scan"] = queryScan
+            outputLine = format_output_line(libraryMetadata, queryMetadata, matchMetadata)
+            outputs.append(outputLine)
+        outputDf = format_output_as_pandas_dataframe(self._queryContext.filePath, outputs)
+        outFile = self._create_outfile_path()
+        outputDf.to_csv(outFile, index=False)
```
Comment (AlexandreHutton) [inline csodiaq/identifier/Identifier.py]: `outFile` should be specified by whatever is calling this method. You can have something like:  
```Python
def _write_identification_to_dataframe(self, matchDf, scoreDf, outFile=None):  
[...]  
if outFile is None:  
  outFile = self._create_outfile_path()  
```
and that should continue to work with the current setup.

File: csodiaq/identifier/matchingFunctions.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+pd.set_option('display.max_columns', None)
+pd.set_option('display.max_rows', None)
+
+def calculate_parts_per_million_relative_difference(referenceMz, targetMz):
+    return (referenceMz - targetMz) * (1e6) / referenceMz
+
+def is_within_tolerance(ppm, tolerance):
+    return abs(ppm) <= tolerance
+
+def match_library_to_query_pooled_spectra(libraryPeaks, queryPeaks, ppmTolerance):
```
Comment (AlexandreHutton) [inline csodiaq/identifier/matchingFunctions.py:9]: I'm inferring that `libraryPeaks` and `queryPeaks` are collections of tuples of form (mZ, intensity, tag). Could you add a docstring stating that?

File: csodiaq/identifier/matchingFunctions.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        p = i + 0
+        while (p < len(libraryPeaks)):
+            ppm = calculate_parts_per_million_relative_difference(libraryPeaks[p][mzIdx], queryPeaks[j][mzIdx])
+            if not is_within_tolerance(ppm, ppmTolerance):
+                break
+            data.append(
+                [libraryPeaks[p][tagIdx], libraryPeaks[p][intensityIdx], queryPeaks[j][tagIdx], queryPeaks[j][intensityIdx], ppm]
+            )
+            p += 1
+        j += 1
```
Comment (AlexandreHutton) [inline csodiaq/identifier/matchingFunctions.py:63]: Consider something like this:
```Python
last_query_idx = 0
for lib_mz, lib_intensity, lib_tag in libraryPeaks:
    for query_mz, query_intensity, query_tag in queryPeaks[last_query_idx:]:
        while is_within_tolerance(calc_ppm(lib_mz, query_mz), ppmTolerance):
            data.append( data_to_append)
        # no longer in tolerance
        if lib_mz > query_mz:
            continue  # go to next element in inner loop
        else:
            last_query_idx += 1  # might need to be moved elsewhere
            break  # go to next element in outer loop
```  
  
I think that should work, but you'll need to check it. That main advantage is that there's no need for loop indices, mz/intensity/tag data are separated without indices, no balancing of i/j/p, and it's more legible.

File: csodiaq/identifier/matchingFunctions.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    return pd.DataFrame(data, columns=["libraryIdx","libraryIntensity","queryIdx","queryIntensity","ppmDifference"])
+
+def eliminate_low_count_matches(matches):
+    minNumMatches = 3
```
Comment (AlexandreHutton) [inline csodiaq/identifier/matchingFunctions.py]:
```Python
def eliminate_low_count_matches(matches, minNumMatches=3):
```

File: csodiaq/identifier/outputWritingFunctions.py (hunk 1 above)
Comment (AlexandreHutton) [reply csodiaq/identifier/outputWritingFunctions.py]: You can select multiple columns by specifying a list for the columns:  
```Python  
outputDict = df.set_index(["libraryIdx", "queryIdx"])[["maccScore", "cosineScore"]].to_dict()  
```

File: csodiaq/identifier/scoringFunctions.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+                    calculate_cosine_similarity_score(x["libraryIntensity"], x["queryIntensity"]),
+                    calculate_macc_score(x["libraryIntensity"], x["queryIntensity"])
+                ))\
+                .reset_index(name="scores")
+    scoreDf["cosineScore"],scoreDf["maccScore"] = zip(*scoreDf["scores"])
+    del scoreDf["scores"]
+    return scoreDf.sort_values("maccScore", ascending=False).reset_index(drop=True)
+
+def calculate_cosine_similarity_score(vectorA, vectorB):
+    return 1 - cosine(vectorA, vectorB)
```
Comment (AlexandreHutton) [inline csodiaq/identifier/scoringFunctions.py:19]: Alternatively, import from [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html).

File: csodiaq/identifier/scoringFunctions.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+def identify_all_decoys(decoySet, scoreDf):
+    return np.where(scoreDf["libraryIdx"].isin(decoySet), 1, 0)
+
+def determine_index_of_fdr_cutoff(isDecoyArray):
+    if isDecoyArray[0]:
+        raise ValueError("None of the library peptides were identified in the query spectra (highest score was a decoy).")
+    if 1 not in isDecoyArray:
+        return len(isDecoyArray)
+    fdrCutoff = 1e-2
```
Comment (AlexandreHutton) [inline csodiaq/identifier/scoringFunctions.py]:
```Python
def determine_index_of_fdr_cutoff(isDecoyArray, fdrCutoff=1e-2):
```

File: csodiaq/identifier/scoringFunctions.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+def calculate_ppm_offset_tolerance(ppms, numStandardDeviations):
+    if numStandardDeviations: return calculate_ppm_offset_tolerance_using_mean_and_standard_deviation(ppms, numStandardDeviations)
+    else: return calculate_ppm_offset_tolerance_using_tallest_bin_peak(ppms)
+
+def calculate_ppm_offset_tolerance_using_mean_and_standard_deviation(ppms, numStandardDeviations):
+    return np.mean(ppms), np.std(ppms) * numStandardDeviations
+
+def create_standardized_histogram(ppms):
+    numBins = 200
```
Comment (AlexandreHutton) [inline csodiaq/identifier/scoringFunctions.py]: Add as default value.

**Review by AlexandreHutton — CHANGES_REQUESTED** [review]
There are a few functions that have hard-coded default values for variables that could be moved to the argument list with no impact on existing code while making them more flexible for future use.  
There are a few bits with recommended changes, particularly `matchingFunctions.py`.  
You can keep the files as-is, but file naming convention in Python is to have them be lower case. A lot of your classes/files have colliding names. You'd normally avoid that by having the file called `_name.py`, having the class/function named `name()`, and using `__init__` to import `name` to that level. Basically pretending that `_name.py` doesn't exist.  
  
I skipped looking at the tests, but I can go back and look over them if you'd like me to; let me know.

File: csodiaq/identifier/Identifier.py (hunk 7 above)
Comment (CCranney) [reply csodiaq/identifier/Identifier.py]: the 'args' parameter is the input from argsparse, which contains commands/flags issued by the user from the command line. Are you just recommending I change the name to "commandLineArgs" or some such to avoid the standard use of 'args' as a function parameter in python?

File: csodiaq/identifier/Identifier.py (hunk 8 above)
Comment (CCranney) [reply csodiaq/identifier/Identifier.py]: are you thinking I should return the dataframe as an output value and do the writing at a higher level, like in csodiaq.py?

File: csodiaq/identifier/Identifier.py (hunk 10 above)
Comment (CCranney) [reply csodiaq/identifier/Identifier.py]: Changed in my next commit

File: csodiaq/identifier/matchingFunctions.py (hunk 12 above)
Comment (CCranney) [reply csodiaq/identifier/matchingFunctions.py:9]: Added, will be in my next commit.

File: csodiaq/identifier/matchingFunctions.py (hunk 14 above)
Comment (CCranney) [reply csodiaq/identifier/matchingFunctions.py]: Added, will be in my next commit.

File: csodiaq/identifier/scoringFunctions.py (hunk 15 above)
Comment (CCranney) [reply csodiaq/identifier/scoringFunctions.py:19]: I played around with doing this, but the formatting and outputs were not as straightforward as desired. It produces a predictive matrix rather than a score. It's possible I could tweak this further to make it work in such a way, but I don't think it would look as clean.

File: csodiaq/identifier/Identifier.py (hunk 7 above)
Comment (CCranney) [reply csodiaq/identifier/Identifier.py]: We talked, I'll just rename the parameter

File: csodiaq/identifier/Identifier.py (hunk 8 above)
Comment (CCranney) [reply csodiaq/identifier/Identifier.py]: Confirmed - yes this would be best.

File: csodiaq/identifier/matchingFunctions.py (hunk 13 above)
Comment (CCranney) [reply csodiaq/identifier/matchingFunctions.py:63]: I've added this as a task to complete in the future, because I don't think I can do it right away.
